ksql-engine/src/test/java/io/confluent/ksql/function/UdfLoaderTest.java

```diff
@@ -572,6 +571,595 @@ public void shouldEnsureFunctionReturnTypeIsDeepOptional() {
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithLongValTypeDoubleAggType() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createLongDouble"),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithDoubleValTypeLongAggType() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createDoubleLong"),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithIntegerValTypeStringAggType() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createIntegerString"),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithStringValTypeIntegerAggType() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createStringInteger"),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithBooleanValTypeListAggType() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createBooleanList"),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithListValTypeBooleanAggType() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createListBoolean"),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithMapValMapAggTypes() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createMapMap"),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithMapValMapAggTypesAndFactoryArg() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createMapMap", int.class),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithStructStructTypes() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createStructStruct"),
+        "test",
+        "desc",
+        "STRUCT<A VARCHAR>",
+        "STRUCT<B VARCHAR>",
+        "STRUCT<B VARCHAR>");
+  }
+
+  @Test(expected = KsqlException.class)
+  public void shouldThrowWhenTryingToGenerateUdafThatHasIncorrectTypes() throws Exception {
+    createUdfLoader().createUdafFactoryInvoker(UdfLoaderTest.class.getMethod("createBad"),
+        "test",
+        "desc",
+        "",
+        "",
+        "");
+  }
+
+  @Test(expected = KsqlException.class)
```

Comment:
For `KsqlException`, where the error message will ultimately be customer facing, it's generally a good idea to add details of the expected exception into the test case, e.g. 

```java
    @Rule
    public final ExpectedException expectedException = ExpectedException.none();

    @Test
    public void shouldThrowWhenUdafFactoryMethodIsntStatic() throws Exception {
    // Given:
    final Method method = UdfLoaderTest.class.getMethod("createNonStatic");

    // Then:
    expectedException.expect(KsqlException.class);
    expectedException.expectMessage("Method FooBar in UDF blah, jar blah, must be static");

    // When:
    createUdfLoader().createUdafFactoryInvoker(
         method,
        "test",
        "desc",
        "",
        "",
        "");
  }
```

Same for other tests that throw.

Reply:
I didn't write these tests, they're copy and pasted from UdfCompilerTest which doesn't exist any more

Reply:
done